dataframe.py

```diff
@@ -0,0 +1,269 @@
+        """
+
+    @abstractmethod
+    def equals(self, other: 'DataType') -> bool:
+        """
+        Return true if other DataType contains the same metadata as this
+        DataType
+        """
+        pass
+
+
+class PrimitiveType(DataType):
+
+    def equals(self, other: DataType) -> bool:
+        return type(self) == type(other)
+
+
+class NullType(PrimitiveType):
+    """
+    A data type whose values are always null
+    """
+    def to_string(self):
+        return "null"
+
+
+class Boolean(PrimitiveType):
+
+    def to_string(self):
+        return "bool"
+
+
+class NumberType(PrimitiveType):
+    pass
+
+
+class IntegerType(NumberType):
+    pass
+
+
+class SignedIntegerType(IntegerType):
+    pass
+
+
+class Int8(SignedIntegerType):
+
+    def to_string(self):
+        return "int8"
+
+
+class Int16(SignedIntegerType):
+
+    def to_string(self):
+        return "int16"
+
+
+class Int32(SignedIntegerType):
+
+    def to_string(self):
+        return "int32"
+
+
+class Int64(SignedIntegerType):
+
+    def to_string(self):
+        return "int64"
+
+
+class Binary(PrimitiveType):
+    """
+    A variable-size binary (bytes) value
+    """
+    def to_string(self):
+        return "binary"
+
+
+class String(PrimitiveType):
+    """
+    A UTF8-encoded string value
+    """
+    def to_string(self):
+        return "string"
+
+
+class Object(PrimitiveType):
+    """
+    Any PyObject value
+    """
+    def to_string(self):
+        return "object"
+
+
+class Categorical(DataType):
+    """
+    A categorical value is an ordinal (integer) value that references a
+    sequence of category values of an arbitrary data type
+    """
+
+    def __init__(self, index_type: IntegerType, category_type: DataType,
+                 ordered: bool = False):
+        self.index_type = index_type
+        self.category_type = category_type
+        self.ordered = ordered
+
+    def equals(self, other: DataType) -> bool:
+        return (isinstance(other, Categorical) and
+                self.index_type == other.index_type and
+                self.category_type == other.category_type and
+                self.ordered == other.ordered)
+
+    def to_string(self):
+        return ("categorical(indices={}, categories={}, ordered={})"
+                .format(str(self.index_type), str(self.category_type),
+                        self.ordered))
+
+
+# ----------------------------------------------------------------------
+# Classes representing a column in a DataFrame
+
+
+class Column(ABC):
+
+    @property
+    @abstractmethod
+    def name(self) -> Hashable:
+        pass
+
+    @property
+    @abstractmethod
+    def type(self) -> DataType:
+        """
+        Return the logical type of each column cell value
+        """
+        pass
+
+    @property
+    def attrs(self) -> Mapping:
+        """
+        Metadata for this column. Default implementation returns empty dict
+        """
+        return {}
+
```

Comment:
Proposing an additional API here for a more generic array API. I think that the constraint of supporting `__array_function__` protocol helps to prevent this from becoming too overloaded, but wanted to kickoff discussion
```suggestion

def to_array(self):
    """
    Access column's data as an array object which supports the
    `__array_function__` protocol. Recommended to return a view if able but
    not required
    """
    raise NotImplementedError("Conversion to array not available")

```

Reply:
Will open another suggestion around `to_dict_of_array` once this is settled.

Reply:
Alright. From my perspective a critical matter is making sure that this doesn't get mixed up with APIs whose purpose is to provide data in a particular memory representation. So I think the term "array" in this context is may confuse, so `to_array_function_protocol` or something more explicit would be clearer (and `__array_function__` can be available also which calls this function).

Reply:
>  From my perspective a critical matter is making sure that this doesn't get mixed up with APIs whose purpose is to provide data in a particular memory representation.

Agreed.

> So I think the term "array" in this context is may confuse, so to_array_function_protocol or something more explicit would be clearer (and __array_function__ can be available also which calls this function).

`to_array_function_protocol_object` is probably the most explicit because we're exporting an object that supports the protocol as opposed to the protocol implementation.

Reply:
> *`to_array_function_protocol_object` is probably the most explicit*

That's... incredibly cumbersome :(

If `to_array` is too ambiguous how about `to_arraylike`

Reply:
> That's... incredibly cumbersome :(

I 100% agree and would be in support of `to_array` or `to_array_like` or something similar, but it does express a form of ambiguity when it comes to expectations for data layout vs API.

Reply:
I think this is where type hits can help - have short names but show the return type:
```python
def to_array(self) -> ArrayLike:
    ...
```

Reply:
Since these is are internal APIs intended for library developers, I think that being explicit and unambiguous is the most important thing. If you prefer a different API, you can create your own syntactic sugar layer that suits your needs.

Reply:
The reason that I'm against `to_array` is then you have this situation

* `to_numpy` -> does data export to a concrete memory representation
* `to_arrow` -> does data export to a concrete memory representation
* `to_array` or `to_array_like` -> what does this do? Based on what I see above it's not doing data export

What is "array" and what is "array_like"? It's too ambiguous. 

IMHO this project is first and foremost about data export at library boundaries (without one library having knowledge of the internals of the other). If you want to also expose object protocols (like `__array_function__`) that's fine, but that needs to not be commingled with the data export APIs.

Reply:
`to_array_like` could export either a numpy array *or* a pyarrow array, or anything which implemented the array protocol. A consumer of this method could then write a generic function, agnostic to whatever the concrete representation was - so long as they stuck to methods implemented in the array protocol.

Reply:
I think this goes beyond the scope of what problem is trying to be solved in this first projecf iteration. There is exporting "data" and exporting "interfaces". If we can't first agree about exporting "data" then I think this project will be kaputt and we should all disengage.

Reply:
> * `to_numpy` -> does data export to a concrete memory representation

From my perspective, in addition to exporting to a concrete memory representation this takes a step too far and exports to a specific object. It would be great if we could generalize this to exporting any object that follows a concrete memory representation as opposed to tightly confining it to a specific object. I.E. (ignore the cumbersome names for now)
- `to_array_interface_object` -> does data export to an object that supports `__array_interface__` protocol
- `to_buffer_protocol_object` -> does data export to an object that exposes Python buffer protocol
- `to_cuda_array_interface_object` -> does data export to an object that supports `__cuda_array_interface__` protocol

These protocols are specifically defined surrounding memory representation so it makes sense to use them from my perspective.

Reply:
I'm reached the limit of how much time I can invest in this for now so I'll wait a few weeks or a month for others to express their opinions. For me the things I care most about are:

* Clear and unambiguous names (even if they are "cumbersome")
* Clear distinction between methods that yield objects implementing some other protocol / interface (like the buffer protocol) and methods that yield particular containers. If a consumer wants a `numpy.ndarray`, I think they should be able to ask for that

Per my comment on the Discourse thread, I do think taking a step back and writing a requirements doc might be helpful to avoid diversions into discussions of non-goals that take up a lot of time and energy.